erlang/src/erl_ast_helper.ml

```diff
@@ -145,6 +145,66 @@ module Type = struct
   let fun_ ?(args = []) ~return =
     Type_function { tyfun_args = args; tyfun_return = return }
 
+  let rec all_named_vars typ_expr =
+    match typ_expr with
+    | Type_const _ -> []
+    | Type_variable (Var_name n) -> [ n ]
+    | Type_variable _ -> []
+    | Type_constr { tc_name = _; tc_args = args } ->
+        List.concat_map all_named_vars args
+    | Type_list p -> all_named_vars p
+    | Type_tuple parts -> List.concat_map all_named_vars parts
+    | Type_map fields -> List.concat_map all_named_vars_map_field fields
+    | Type_record (_, fields) ->
+        List.concat_map all_named_vars_record_field fields
+    | Type_variant tks -> List.concat_map all_named_vars tks
+    | Type_function { tyfun_args = args; tyfun_return = return } ->
+        List.concat_map all_named_vars (return :: args)
+
+  and all_named_vars_map_field
+      { tmf_name = e1; tmf_presence = _; tmf_value = e2 } =
+    List.concat_map all_named_vars [ e1; e2 ]
+
+  and all_named_vars_record_field { rf_name = _; rf_type = t } =
+    all_named_vars t
+
+  let rec clean_vars occurs typ_expr =
+    match typ_expr with
+    | Type_const l -> Type_const l
+    | Type_variable (Var_name n)
+      when List.length (List.find_all (( = ) n) occurs) < 2 ->
+        Type_variable (Var_name "_")
```

Comment:
This is the main logic. Probably, it's worth to extract the condition into a separate function something like `is_var_unbound`?

Reply:
Makes sense, we can do this small refactors later on.